src/components/notes/saving-indicator.jsx

```diff
@@ -19,6 +20,61 @@ const outOfSyncAfter = 30 * sec
  * @property {Ref} bannerRef - Where to display banners
  */
 
+/**
+ * Checks if the server is out of sync
+ * (our changes were not successfully sent and accepted for too much time)
+ * @param {object} params
+ * @param {boolean} params.isDirty - if there are changes to be sent
+ * @param {integer} params.dirtyAge - number of milliseconds waiting
+ *                                    with changes to be sent
+ * @returns {boolean}
+ */
+function checksOutOfSync({ isDirty, dirtyAge }) {
+  return isDirty && dirtyAge && dirtyAge > outOfSyncAfter
+}
+
+/**
+ * Checks for disconnection (browser offline or out of sync)
+ *
+ * @param {object} params
+ * @param {boolean} params.isOutOfSync
+ * @returns {{isDisconnected, disconnectedAge}}
+ */
+function checksDisconnection({ isOutOfSync }) {
+  const now = new Date()
+  const isBrowserOffline = useBrowserOffline()
+  const [disconnectedSince, setDisconnectedSince] = useState()
+  const isDisconnected = !!disconnectedSince
+  useEffect(() => {
+    const disconnectedNow = isBrowserOffline || isOutOfSync
+    if (!!disconnectedNow !== !!isDisconnected) {
+      setDisconnectedSince(disconnectedNow ? new Date() : null)
+    }
+  }, [isBrowserOffline, isOutOfSync, setDisconnectedSince, isDisconnected])
+  const disconnectedAge = isDisconnected ? now - disconnectedSince : 0
+  return { isDisconnected, disconnectedAge }
+}
```

Comment:
Nit: I don't understand why we have isOutOfSync in this function ? Do you think it could be "mixed" in later by the caller ? I think having both concepts "offline" and "out of sync" is a bit confusing.

Reply:
I use the following vocabulary : 

- Offline: the browser is offline, or declares to be offline
- Out of sync: for an unknown reason (no connection, error on the server, or anything else), we couldn't commit our change for a long time
- Disconnected: We know we are not correctly connected to the server (either offline or out of sync)

This hook/function is for the third case, and we need it to remember the date in which we enter this state

Reply:
Nit: I wonder if this "remember-the-date-in-which-we-enter-this-state" could not be expressed with a dedicated hook ?

```
const useAge = value => {
  const [lastValue, setLastValue] = useState(value)
  const [lastTime, setLastTime] = useState(Date.now())
  useEffect(() => {
    if (value !== lastValue) {
      setLastTime(Date.now())
      setLastValue(value)
    }
  }, [value, lastValue, setLastValue])
  return lastTime
}

const useDisconnection = ({ isOutOfSync }) => {
  const isBrowserOffline = useBrowserOffline()
  const isDisconnected = isBrowserOffline || isOutOfSync
  const disconnectedValueAge = useAge(isDisconnected)
  const disconnectedAge = isDisconnected ? disconnectedValueAge : 0
  return { isDisconnected, disconnectedAge }
}
```

Reply:
I was doing that this morning, and found something way simplier which probably doesn't need any helper : 

```
const disconnectedSince = useMemo(() => new Date(), [isDisconnected])
```
which gives
```
function useDisconnection({ isOutOfSync }) {
  const now = new Date()
  const isBrowserOffline = useBrowserOffline()
  const isDisconnected = isBrowserOffline || isOutOfSync
  const disconnectedSince = useMemo(() => new Date(), [isDisconnected])
  const disconnectedAge = isDisconnected ? now - disconnectedSince : 0
  return { isDisconnected, disconnectedAge }
}
```